Point: redefine distance() in Point to work for Line and other GeometryEntity

friyaz opened the pull request:
point : modified `distance` to work for any Linear Entity . 

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #15531 


#### Brief description of what is fixed or changed
`distance()` is made symmetric . 
For objects of type `Point` , `tuple` or `list` , distance is calculated using distance formula directly by converting into Point if object is not already a Point . A `ValueError` is raised if any other non `GeometryEntity` is given as argument. If` GeometryEntity` has attribute distance then distance is calculated using 
`other.distance(self)`
Otherwise `Attribute Error` is raised. 

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* geometry
   * `distance` when called from object of Point type can return distance from line . 
<!-- END RELEASE NOTES -->

jksuom commented on sympy/geometry/point.py:
```diff
@@ -402,18 +403,23 @@ def distance(self, p):
-        s, p = Point._normalize_dimension(self, Point(p))
+        from sympy.geometry.line import LinearEntity
+        if isinstance(other , LinearEntity) :
+            other = other.projection(self)
```
How about using `return other.distance(self)`?

FrackeR011 replied:
I was thinking about that too but i was not sure which way to choose.
So i just decided to go with this approach.
` return other.distance(self) ` will not work for objects of other class which inherit from 'LinearEntity` like `ray` and `segment` .
But `distance` is not defined in class`ray` and `segment` so i believe this may never be needed .
 I argee with you that this approach is better .

FrackeR011 replied:
I need to check whether the `projection` of point p lie within the segment for distance between line and segment to make sense.

normalhuman replied:
What if other is a Plane? I think a more general, future-proof approach would be to 

1. If `other` is of a type that can be handled here (Point) handle it.
2. Try `other.distance(self)`, but catch and re-throw any AttributeError or TypeError to give a more informative error message.   

This makes evaluation is symmetric and eliminates hard-coding class references like LinearEntity.

FrackeR011 replied:
By mistake i have deleted my previous commit .Sorry for that.

friyaz committed:
Point : made distance symmetric

    distance function first checks whether argument passed( other) is a Geomety entity.
    If other is not a GeometryEntity , it is converted into a Point if it is a tuple or a list
    otherwise ValueError is raised .
    Further Ifit is a  point then distance is directly calculated using distance formula.
    If it is a any other GeometryEntity such that other.distance is defined it is returned
    otherwise AttributeError is raised.

    Test case is added.

jksuom commented on sympy/geometry/point.py:
```diff
@@ -380,19 +380,23 @@ def are_coplanar(cls, *points):
         ==========
 
-        p : Point
+        p : LinearEntity or Point
```
There could probably be GeometricEntity. Distance from non-linear geometric objects like a circle might be later implemented.

FrackeR011 replied:
Actually `p` is reference to the point object after line
`s, p = Point._normalize_dimension(self, Point(other))`
It should be `point` only.
`p` does not refer to `other` which can either be a GeometricEntity or LinearEntity.

jksuom commented on sympy/geometry/point.py:
```diff
@@ -402,19 +406,28 @@ def distance(self, p):
-        s, p = Point._normalize_dimension(self, Point(p))
-        return sqrt(Add(*((a - b)**2 for a, b in zip(s, p))))
+        if not isinstance(other , GeometryEntity) :
+            other = Point(other, dim=self.ambient_dimension)
```
Maybe a TypeError should be raised. Or else docstring should be extended.

friyaz committed:
improved docstringof distance

friyaz committed:
Modified TypeError message in distance
TypeError is raised when Non GeometricEntity other than list or tuple is passed.

smichr commented on sympy/geometry/point.py:
```diff
@@ -402,19 +404,31 @@ def distance(self, p):
+            try :
+                other = Point(other, dim=self.ambient_dimension)
+            except TypeError :
+                raise TypeError("only distnace with some GeometricEntity, Point, List or tuple is supported")
```
distnace->distance

smichr committed:
Update point.py

smichr committed:
Update point.py

Rather than discuss the specific ways that a Geometric object may be recognized, just indicate that an error is raised if it is not recognized. Also add a comment to the docstring regarding the symbolic distance that is computed.
